### PR TITLE
[precompile] Capture across the caller's loop, and keep input-rooted guards

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #195352
* #195351
* #195664
* #195350
* #195349
* #195348
* #195347
* #195346

precompile makes its example calls itself, back to back, which is correct only
when the calls are independent. A pipelined training step is not: its inputs
come off a queue the enclosing loop advances, so the second call finds the
state the first one consumed, and nothing precompile can do while it drives
the iteration fixes that. That is the case the single entry point in #195348
cannot express, and the reason the caller-driven form it declined to add is
worth having. precompile.accumulate inverts control -- the caller keeps their
loop, and each call runs for real, folds in what it newly exercised, rewrites
both files, and returns what the region returned, so there is no finalize step
and a job that dies partway still leaves a loadable artifact. It is a handle
(AccumulatingCapture, exported from torch.compiler) and not a keyword argument
because the state cannot be a value: variants are filed under an
isolate_recompiles_id that nothing can hand back to torch._dynamo.optimize, so
they are reachable only while the compiled region is alive. close() releases
it. A gate that refuses a render does not lose the call: the result is still
returned, the files keep the previous artifact, the refusal is logged once per
distinct message, and the last one is raised from close() unless a later
render passed. A *args or **kwargs entry is refused before the first call runs,
since on this path the render-time check would otherwise surface at close().
Like load, accumulate reports torch.compiler as its public module.

Re-entering one session per call is the rest of the accumulate work.
PrecompileSession takes a resumable flag fixed at construction: __exit__ reads
it to decide whether the one-shot policy pass runs, so a close() cannot turn a
resumable session into a one-shot one under a call in flight. A call that
raised has already told the caller, so each render's completeness gate judges
only the errors that call produced, rather than freezing the artifact at the
last good batch for the rest of the loop; and the recorded guard leaves
accumulate across cycles instead of being replaced, since
record_live_guard_leaves yields a fresh set on every entry while drift is
reported against every frame in the package, so keeping the last cycle's alone
would flag every frame that call did not recompile. Rendered subgraph source
is cached per backend id, since a backend id's graph never changes and an
accumulating capture renders after every call, and the drift and
unrebuildable-guard warnings are issued once per (frame, leaf set) and once
per (frame, slot) respectively rather than once per render. The handle
serializes calls from several threads under one lock, and close() takes the
same lock, so it waits for a call in flight and the rewrite that call ends in;
the session records the thread that entered it and a close() from another
thread waits for that thread's __exit__ before retiring the region, because
the capture config __enter__ installs is context-local and only that thread
can undo it. That mark is set inside __enter__'s try, so a failure while
entering (the config patch raising, say) clears it on the way out instead of
leaving every other thread's close() waiting for an __exit__ that never
comes. And both waits know who is waiting: the session counts calls in flight
per thread, and a close() or __exit__ from a thread that is inside one of
them raises RuntimeError rather than wait on itself, as does the handle's
close() when fn, or a hook it fires, calls it.

Accumulating forced the question this PR mostly answers: the invariant-guard
policy cannot be applied twice. It rewrites each variant's serialized guards
and then prunes the facts it derived them from, so a second pass filters bytes
the first already filtered against facts the first already edited.
varying_guard_slots over a single variant is empty by construction, so the
pass after call one strips every droppable slot of every frame, and a later
call that makes one of them discriminate cannot give it back.
_apply_guard_policy still consumes the session on the one-shot path, where the
capture is over and the facts are spent; the accumulating render goes through
_policy_filtered_codes, which runs the same core (_run_guard_policy) over deep
copies. That render also runs the policy before its gates, which otherwise
read the previous render's numbers -- inert on the first call, so a capture
that rendered once reported POLICY_DROPPED_GUARDS = [] for a pass that had
dropped every invariant slot.

That one variant cannot disagree with itself is also why the policy's
exemption changes shape. #195348 exempts by name the five guard types that pin
an input's shape or value, and the types that pin a branch, which stops a
wrong-shaped call being answered and keeps a hasattr branch from being served
to the side it did not capture. But a list of types is only as complete as
the list, and the inference underneath is untouched: a slot invariant across
variants is taken for environment, and with one variant that runs on no
evidence, so a slot of any type the list does not name goes too, even when
the caller's own data roots it. Captured with an attribute present and served
without it, that is the captured side's numbers under fully default gates
with RISKY_DROPPED_GUARDS empty, and that failure shape is what the root rule
closes off in general rather than per type. Slots are classified by ROOT --
module structure, process globals, interpreter state and the shape env are
the environment, anything else a local holds is caller data -- and
module-rooted is decided by VALUE rather than by a name beginning "self",
since precompile's convention passes the model in as an argument. It keeps
little extra: on a ranking capture of 7,575 slots, the root rule retains
roughly 60. Each drop is reported with the check it rendered as, because a
type and a source cannot tell a benign HASATTR companion of a kept
TENSOR_MATCH from the only thing pinning an optional attribute.

The summary grows the fields the header reports: the policy's drops with their
rendered checks (dropped_guard_code / DROPPED_GUARD_CODE), guards that rebuilt
into a different check (drifted_guards / DRIFTED_GUARDS), variants left with
no input-rooted guard (variants_without_input_guards /
VARIANTS_WITHOUT_INPUT_GUARDS), and variants dropped because inductor's CPU
codegen target moved (variants_dropped_for_codegen_target /
VARIANTS_DROPPED_FOR_CODEGEN_TARGET). That last one is a coverage gap like a
truncated frame -- a call the capture ran and does not serve -- so a nonzero
count makes complete False and the default require_complete refuse, naming
the cause; #195351 reported it and accepted the artifact. kept_guards
subtracts the policy's drops so the two never overlap. The summary fields
carry #: doc comments so autodoc shows them, and the public docs gain
accumulate with its parameters and keep_example_grads.

One thing moves for existing callers: the guard-policy change alters what a
default dynamo capture prunes, and therefore what its POLICY_DROPPED_GUARDS
header and PrecompileSummary.kept_guards report. keep_example_grads is an
addition rather than a change: False is today's snapshot/clear/restore around
the example calls, and True is for the caller whose example call IS their live
training step, where restoring silently discards the backward they just paid
for. accumulate skips the snapshot altogether, since every call is the
caller's own.

Review accumulate() and AccumulatingCapture in torch/_precompile.py and their
docs first, then precompile_package.py -- resumability and close(), the policy
split, root classification -- then the summary fields in
_precompile_types.py.

Test Plan:

```
python test/test_precompile.py
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_precompile.py
lintrunner -a docs/source/torch.compiler_api.md test/test_precompile.py torch/_dynamo/precompile_package.py torch/_precompile.py torch/compiler/__init__.py torch/compiler/_precompile_types.py
```

The guard-policy tests assert the artifact REFUSES the call it cannot answer
rather than answering it wrongly: the hasattr and derived-type (dict length)
cases capture under fully default gates and serve the captured branch's
numbers without this change, while a module-heavy capture must still drop
slots and every slot it drops must be module- or global-rooted, so the root
rule did not buy correctness by keeping everything; the miss message names the
first failing check of an input-rooted guard the policy now keeps, every
dropped slot in the header carries its rendered check, and a frame with no
inputs is counted as having no input guards. Unfixed, the accumulate test
whose second call makes a slot vary answers that call with the first one's
numbers. The accumulate tests also load the artifact after every call, check
the session's own guard state is untouched by a render, survive a call that
raised, return the result when a gate refuses, refuse calls after close(),
refuse make_fx and a lone path, refuse a varargs entry before any call and
accept a torch._dynamo.disable-wrapped one, and let a module fn's gradients
accumulate. The concurrency tests close a capture from another thread while a
call is in flight and check the rendered POLICY_DROPPED_GUARDS is identical to
a sequential run's -- re-raising the worker's exception and asserting the
ordering (the call was counted before close() returned) rather than elapsed
time -- that a session close from a second thread waits for the entering
thread's exit and returns when that entry raised, that a handle shared across
threads serializes its calls, and that close() from inside the call raises
instead of deadlocking. The keep_example_grads tests check the default
restore, eager-identical accumulation with the flag, and that it is refused
for make_fx. test_summary_reports_variants_dropped_for_codegen_target checks
the dropped variant makes complete False and require_complete refuse.
test_precompile_serving_does_not_mutate_the_artifact pins the deep copy
#195347's initialize() takes before the first install, by serving a new shape
per round so each install recompiles and writes back; it needs load() and the
dynamo artifact from later commits, so it lives here. The public-location and
type-hint tests are parametrized over accumulate. The dynamo-wrapped skip list
gains the four tests here the outer trace breaks (two accumulate round trips
and the two policy tests that read the session's own drop set) and no more.

Authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by #195360/#195362 in the canonical precompile stack #195353-#195362. Not carried over: the 'First failing check: variant N' miss text and the warning for variants with only global-state guards; re-file as follow-ups if wanted. Closing.
